File: semester 4/MPP/C#FTS/C#FTS/Repository/DBRepository/ShowRepository.cs
using C_FTS.Domain;
using C_FTS.Utils;
using log4net;
using log4net.Repository.Hierarchy;
using System;
using System.Collections.Generic;
using System.Data;
using System.Text;

namespace C_FTS.Repository.DBRepository
{
    internal class ShowRepository : IShowRepository
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(ShowRepository));
        private Database db = new Database();
        private IArtistRepository artistRepository = new ArtistRepository();
        private IVenueRepository venueRepository = new VenueRepository();

        public void Delete(Show entity)
        {
            throw new NotImplementedException();
        }

        public Show? Find(int id)
        {
            logger.InfoFormat("Entering Find with id {0}", id);
            try
            {
                IDbConnection conn = db.GetConnection();

                using (IDbCommand cmd = conn.CreateCommand())
                {
                    cmd.CommandText = "SELECT * FROM show WHERE id = @id";
                    IDbDataParameter dp = cmd.CreateParameter();
                    dp.ParameterName = "@id";
                    dp.Value = id;
                    cmd.Parameters.Add(dp);

                    using (var dataR = cmd.ExecuteReader())
                    {
                        if (dataR.Read())
                        {
                            Venue venue = venueRepository.Find(dataR.GetInt32(dataR.GetOrdinal("venue_id")));
                            Show show = new Show(dataR, venue);
                            logger.InfoFormat("Exiting Find with {0}", show);
                            return show;
                        }

                        logger.WarnFormat("Find: no Show found with id {0}", id);
                        return null;
                    }
                }
            }
            catch (Exception ex)
            {
                logger.ErrorFormat("Find: exception while finding Show with id {0}: {1}", id, ex.Message);
                throw;
            }
        }

        public List<Show> FindAll()
        {
            logger.Info("Entering FindAll");
            List<Show> shows = new List<Show>();

            try
            {
                IDbConnection conn = db.GetConnection();

                using (IDbCommand cmd = conn.CreateCommand())
                {
                    cmd.CommandText = "SELECT * FROM show";

                    using (var dataR = cmd.ExecuteReader())
                    {
                        while (dataR.Read())
                        {
                            Venue venue = venueRepository.Find(dataR.GetInt32(dataR.GetOrdinal("venue_id")));
                            Show show = new Show(dataR, venue);
                            shows.Add(show);
                        }
                    }
                }

                logger.InfoFormat("FindAll: returning {0} show(s)", shows.Count);
                return shows;
            }
            catch (Exception ex)
            {
                logger.ErrorFormat("FindAll: exception while retrieving all shows: {0}", ex.Message);
                throw;
            }
        }

        public Show? Save(Show entity)
        {
            throw new NotImplementedException();
        }

        public Show? Update(Show entity)
        {
            logger.InfoFormat("Entering Update for Show id {0}", entity.Id);
            try
            {
                IDbConnection conn = db.GetConnection();

                using (IDbCommand cmd = conn.CreateCommand())
                {
                    cmd.CommandText = "UPDATE show SET date = @date, title = @title, sold_seats = @sold_seats, venue_id = @venue_id WHERE id = @id";

                    var dpDate = cmd.CreateParameter();
                    dpDate.ParameterName = "@date";
                    dpDate.Value = entity.Date;

                    var dpTitle = cmd.CreateParameter();
                    dpTitle.ParameterName = "@title";
                    dpTitle.Value = entity.Title;

                    var dpSoldSeats = cmd.CreateParameter();
                    dpSoldSeats.ParameterName = "@sold_seats";
                    dpSoldSeats.Value = entity.SoldSeats;

                    var dpVenue = cmd.CreateParameter();
                    dpVenue.ParameterName = "@venue_id";
                    dpVenue.Value = entity.Venue.Id;

                    var dpId = cmd.CreateParameter();
                    dpId.ParameterName = "@id";
                    dpId.Value = entity.Id;

                    cmd.Parameters.Add(dpDate);
                    cmd.Parameters.Add(dpTitle);
                    cmd.Parameters.Add(dpSoldSeats);
                    cmd.Parameters.Add(dpVenue);
                    cmd.Parameters.Add(dpId);

                    var result = cmd.ExecuteNonQuery();

                    if (result != 0)
                    {
                        logger.InfoFormat("Update: successfully updated Show with id {0}", entity.Id);
                        return entity;
                    }

                    logger.WarnFormat("Update: no rows affected for Show with id {0}", entity.Id);
                    throw new RepositoryException("Error updating show");
                }
            }
            catch (RepositoryException)
            {
                throw;
            }
            catch (Exception ex)
            {
                logger.ErrorFormat("Update: exception while updating Show with id {0}: {1}", entity.Id, ex.Message);
                throw;
            }
        }
    }
}

File: semester 4/MPP/JFTS/FestivalPersistance/src/main/java/ro/mpp/repository/DBRepository/ShowRepository.java
package ro.mpp.repository.DBRepository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp.domain.Show;
import ro.mpp.domain.Venue;
import ro.mpp.repository.IShowRepository;
import ro.mpp.utils.Database;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class ShowRepository implements IShowRepository {
    private final Database db;
    private final VenueRepository venueRepository;

    private static final Logger logger = LogManager.getLogger(ShowRepository.class);

    public ShowRepository(Database database, VenueRepository venueRepository) {
        this.db = database;
        this.venueRepository = venueRepository;
    }

    @Override
    public Optional<Show> save(Show entity) {
        logger.traceEntry("ShowRepository save {}", entity);
        String sql = "INSERT INTO show (date, title, soldSeats, venue_id) VALUES (?, ?, ?, ?)";

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {

                ps.setString(1, entity.getDate());
                ps.setString(2, entity.getTitle());
                ps.setInt(3, entity.getSoldSeats());
                ps.setInt(4, entity.getVenue().getId());

                ps.executeUpdate();

                try (ResultSet rs = ps.getGeneratedKeys()) {
                    if (rs.next()) {
                        entity.setId(rs.getInt(1));
                    }
                }
                return Optional.of(entity);
            }
        } catch (SQLException e) {
            logger.error("Error while saving show", e);
            throw new RuntimeException("Error while saving show: " + e.getMessage());
        }
    }

    @Override
    public Optional<Show> update(Show entity) {
        logger.traceEntry("ShowRepository update {}", entity);
        String sql = "UPDATE show SET date = ?, title = ?, sold_seats = ?, venue_id = ? WHERE id = ?";

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql)) {

                ps.setString(1, entity.getDate());
                ps.setString(2, entity.getTitle());
                ps.setInt(3, entity.getSoldSeats());
                ps.setInt(4, entity.getVenue().getId());
                ps.setInt(5, entity.getId());

                ps.executeUpdate();

                return Optional.of(entity);

            }
        } catch (SQLException e) {
            logger.error("Error while updating show", e);
            throw new RuntimeException("Error while updating show: " + e.getMessage());
        }
    }

    @Override
    public void delete(Show entity) {
        logger.traceEntry("ShowRepository delete {}", entity);

        String deleteJoin = "DELETE FROM show_artist WHERE show_id = ?";
        String deleteShow = "DELETE FROM show WHERE id = ?";

        try {
            Connection conn = db.getConnection();
            logger.info("ShowRepository deleting from join table for {}", entity);
            try (PreparedStatement ps = conn.prepareStatement(deleteJoin)) {
                ps.setInt(1, entity.getId());
                ps.executeUpdate();
            }

            logger.info("ShowRepository deleting {}", entity);
            try (PreparedStatement ps = conn.prepareStatement(deleteShow)) {
                ps.setInt(1, entity.getId());
                ps.executeUpdate();
            }

        } catch (SQLException e) {
            logger.error("Error while deleting show", e);
            throw new RuntimeException("Error while deleting show: " + e.getMessage());
        }
    }

    @Override
    public Optional<Show> find(Integer id) {
        logger.traceEntry("ShowRepository find show with id {}", id);
        String sql = "SELECT * FROM show WHERE id = ?";

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setInt(1, id);

                try (ResultSet rs = ps.executeQuery()) {
                    if (rs.next()) {
                        Venue venue = venueRepository.find(rs.getInt("venue_id")).orElse(null);
                        return Optional.of(new Show(rs, venue));
                    }
                }

            }
        } catch (SQLException e) {
            logger.error("Error while finding show with id {}", id, e);
            throw new RuntimeException("Error while finding show: " + e.getMessage());
        }
        return Optional.empty();
    }

    @Override
    public List<Show> findAll() {
        logger.traceEntry("ShowRepository findAll");

        String sql = "SELECT * FROM show";
        List<Show> shows = new ArrayList<>();

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql)) {

                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) {
                        Integer showId = rs.getInt("id");
                        Venue venue = venueRepository.find(rs.getInt("venue_id")).orElse(null);
                        shows.add(new Show(rs, venue));
                    }
                }

                return shows;
            }
        } catch (SQLException e) {
            logger.error("Error while finding all shows", e);
            throw new RuntimeException("Error while finding all shows: " + e.getMessage());
        }
    }
}